### ISSUE TITLE
[BUG] @NonNull on primitives arrays does not compile in Eclipse (Java 7)

### ISSUE THREAD

**Issue by phcouton** [rzwitserloot/lombok#2019]
Using Lombok 1.18.4, the following code :
```java
  public void foo(@NonNull byte[] data) {
    // ...
  }
```
compiles with Java 7's javac, but not inside Eclipse. This occurs only for arrays of primitives: (`@NonNull String[] data` works), and this problem does not occurs with Lombok 1.18.2.

To be clear, my goal is to guarantee that `data` won't be null.

I have opened this [Eclipse ticket](https://bugs.eclipse.org/bugs/show_bug.cgi?id=543304). The answers I got makes me perplex :
* the proposed syntax `byte @NonNull [] data` won't compile in Java 7, so I can't use it
* It seems that this problem is triggered by the introduction of `ElementType.TYPE_USE` in the annotation.

Eclipse guys seems to say that TYPE_USE is incorrectly used in Lombok annotation, see in particular [this comment](https://bugs.eclipse.org/bugs/show_bug.cgi?id=543304#c6) and [this one](https://bugs.eclipse.org/bugs/show_bug.cgi?id=543304#c11).

This could be related to this commit: https://github.com/rzwitserloot/lombok/commit/d7873f2d21564e8e7f22409fe03681d7dd4c8c1e

**Comment by rzwitserloot** [comment]
Okay, well, eclipse/Stephen Habermann misanalysed. Lombok is (mostly) fine here.

I've dumped a very detailed analysis of this over on the eclipse bug (the ticket you opened), because this is something eclipse needs to fix, we can't help without causing more pain than we solve.

There is one thing where lombok can be considered to be wrong. Given this code: `void foo(@NonNull String[] arg)`, then:

 in java 1.6/1.7 TYPE_USE should be ignored entirely; this is an annotation on the parameter and that is that. Therefore lombok should generate a nullcheck in the method. Lombok does just that. Note, javac also interprets things this way, whether you use javac1.7, or supply `-source 1.7 -target 1.7` to javac 1.8, 9, 10, or 11. You can confirm this by checking the class file with the `javap -v` command.

in java 1.8+, the annotation can be interpreted to be on the entire parameter, as well as on the `String` part of `String[]`. The spec states that both interpretations are to be treated as correct simultaneously. This weird state of affairs means that this one annotation counts twice. This is in fact precisely what both javac and ecj in 1.8+ mode do: This one annotation is visible twice in 2 different locations when inspecting the produced classfile with `javap -v`. There is only one obvious interpretation of this dual state of affairs: Both the array, as well as each element inside of it, is to be considered as 'should never be null'. Lombok has no support for doing nullchecks in array _elements_, and therefore should only do what lombok does support, which is: Generate a nullcheck on the array. Fortunately this means the behaviour of lombok does not change even though the interpretation of the annotation did, between java7 and java8.

As far as eclipse is concerned, as I advised in the eclipse ticket, on j1.7- compliance mode they should not complain. With j1.8+ compliance mode it is feasible to state that `@NonNull byte[]` is a nullity issue because you can't mark `byte` as `@NonNull`, in which case you can (because you're in 1.8+ compliance mode) turn that into `byte @NonNull []` instead, at which point eclipse will stop complaining.

This leads us into the one bug that lombok does suffer from right now: If you type this: `void test(String @lombok.NonNull arg)`, lombok generates no null check, and it really should.

I'll address that and then close this issue. If eclipse elects to do the wrong thing here and disregard this bug report there's really nothing we can do about it unfortunately.

**Comment by rspilker** [comment]
We're still working on it.

**Comment by rspilker** [comment]
@rzwitserloot See also 1999, which is probably related.

**Comment by rzwitserloot** [comment]
2 hours of research in the JLS, typed half an essay with the fine folks over at eclipse about the intricacies of non-null annotations (the sheer complexity of that topic continues to boggle the mind), 10 hours of coding, including adding a new type of annotation handling which meant I had to teach 10 year old code, and all for a bug that's so obscure we've decided not to list it in the changelog.

Nevertheless, that was an awesome ride and I learned a lot. Glad it's done, and thanks for reporting it!
